### ISSUE TITLE
nyc doesnt cover all paths when using rewire@3

### ISSUE THREAD

**Issue by bforbis** [istanbuljs/nyc#747]
## Expected Behavior
`nyc` should cover the same code paths as `istanbul`

## Observed Behavior
When in a project that has a dependency on `rewire@3`, `nyc` will sometimes not report all the covered code that gets executed. This behavior does not occur when using `rewire@2` or when using `istanbul@0.4.5`

## Bonus Points! Code (or Repository) that Reproduces Issue
I am not entirely clear on what exactly causes this issue. It doesn't happen all the time that rewire is being used, and load order of the files seems to matter.

Below is a repository I created that reproduces the issue in a relatively small number of files.
https://github.com/bforbis/istanbul-rewire-test

## Forensic Information

**Operating System:** macOS Sierra
**Environment Information:**
- node version: 8.9.1
- npm version: 5.5.1

**Comment by skygragon** [comment]
Have the same issue, the coverage report is totally different if using rewire@3.0.2

**Comment by bcoe** [comment]
@skygragon @bforbis what immediately jumps out at me is that `rewire@3` uses `babel`, whereas `rewire@2` does not appear to use babel. I've bumped into issues in the past where the order of babel transpilations can break Istanbul.

I think it might also be worth opening a tracking issue on the `rewire` repo -- my guess is that there's an incompatibility between the two libraries that will need to be resolved (it will probably be somewhat annoying to hunt down).

**Comment by brocoli** [comment]
Right now this is the only thing preventing us from using async/await on our projects..

**Comment by bforbis** [comment]
`rewire@4` has just been published which seems to fix this issue by removing its dependency on `babel-core`.

There was one issue I did run into while upgrading one of my repositories to use `rewire@4` which caused the following failure:
```
"before all" hook:
     Error: global leak detected: __core-js_shared__
```
If using `nyc` in cache mode, this will fail the first time and pass the second time while reading from the cache. However, I am unable to use caching because my testing is done in a clean environment within a CI/CD pipeline.

I do not have a minimum reproduction to share that shows this, but I found that updating my mocha command to `mocha --require babel-register` fixes this issue.
